File: example/src/main/java/com/github/j5ik2o/pekko/persistence/effector/example/javaimpl/BankAccountCommand.java
package com.github.j5ik2o.pekko.persistence.effector.example.javaimpl;

import java.io.Serializable;
import org.apache.pekko.actor.typed.ActorRef;

/** Sealed interface representing bank account commands */
public sealed interface BankAccountCommand extends Serializable {
  /**
   * Get the aggregate ID
   *
   * @return Aggregate ID
   */
  BankAccountId getAggregateId();

  /**
   * Get balance command
   *
   * @param aggregateId Aggregate ID
   * @param replyTo Reply destination
   */
  record GetBalance(BankAccountId aggregateId, ActorRef<CommandReply.GetBalanceReply> replyTo)
      implements BankAccountCommand {
    @Override
    public BankAccountId getAggregateId() {
      return aggregateId;
    }
  }

  /**
   * Stop command
   *
   * @param aggregateId Aggregate ID
   * @param replyTo Reply destination
   */
  record Stop(BankAccountId aggregateId, ActorRef<CommandReply.StopReply> replyTo)
      implements BankAccountCommand {
    @Override
    public BankAccountId getAggregateId() {
      return aggregateId;
    }
  }

  /**
   * Create command
   *
   * @param aggregateId Aggregate ID
   * @param replyTo Reply destination
   */
  record Create(BankAccountId aggregateId, ActorRef<CommandReply.CreateReply> replyTo)
      implements BankAccountCommand {
    @Override
    public BankAccountId getAggregateId() {
      return aggregateId;
    }
  }

  /**
   * Deposit cash command
   *
   * @param aggregateId Aggregate ID
   * @param amount Amount
   * @param replyTo Reply destination
   */
  record DepositCash(
      BankAccountId aggregateId, Money amount, ActorRef<CommandReply.DepositCashReply> replyTo)
      implements BankAccountCommand {
    @Override
    public BankAccountId getAggregateId() {
      return aggregateId;
    }
  }

  /**
   * Withdraw cash command
   *
   * @param aggregateId Aggregate ID
   * @param amount Amount
   * @param replyTo Reply destination
   */
  record WithdrawCash(
      BankAccountId aggregateId, Money amount, ActorRef<CommandReply.WithdrawCashReply> replyTo)
      implements BankAccountCommand {
    @Override
    public BankAccountId getAggregateId() {
      return aggregateId;
    }
  }
}

File: example/src/main/scala/com/github/j5ik2o/pekko/persistence/effector/example/scalaimpl/BankAccountCommand.scala
package com.github.j5ik2o.pekko.persistence.effector.example.scalaimpl

import com.github.j5ik2o.pekko.persistence.effector.example.*
import com.github.j5ik2o.pekko.persistence.effector.example.scalaimpl.*
import org.apache.pekko.actor.typed.ActorRef

enum BankAccountCommand {
  case GetBalance(override val aggregateId: BankAccountId, replyTo: ActorRef[GetBalanceReply])
  case Stop(override val aggregateId: BankAccountId, replyTo: ActorRef[StopReply])
  case Create(
    override val aggregateId: BankAccountId,
    limit: Money = Money(100000, Money.JPY),
    balance: Money = Money(0, Money.JPY),
    replyTo: ActorRef[CreateReply])
  case DepositCash(override val aggregateId: BankAccountId, amount: Money, replyTo: ActorRef[DepositCashReply])
  case WithdrawCash(override val aggregateId: BankAccountId, amount: Money, replyTo: ActorRef[WithdrawCashReply])

  def aggregateId: BankAccountId = this match {
    case GetBalance(aggregateId, _) => aggregateId
    case Stop(aggregateId, _) => aggregateId
    case Create(aggregateId, _, _, _) => aggregateId
    case DepositCash(aggregateId, _, _) => aggregateId
    case WithdrawCash(aggregateId, _, _) => aggregateId
  }
}

object BankAccountCommand {
  type Message = BankAccountCommand
}

enum StopReply {
  case Succeeded(aggregateId: BankAccountId)
}

enum GetBalanceReply {
  case Succeeded(aggregateId: BankAccountId, balance: Money)
}

enum CreateReply {
  case Succeeded(aggregateId: BankAccountId)
}

enum DepositCashReply {
  case Succeeded(aggregateId: BankAccountId, amount: Money)
  case Failed(aggregateId: BankAccountId, error: BankAccountError)
}

enum WithdrawCashReply {
  case Succeeded(aggregateId: BankAccountId, amount: Money)
  case Failed(aggregateId: BankAccountId, error: BankAccountError)
}
